cluster-creation-cli/README.md

```diff
@@ -0,0 +1,163 @@
+
+- A **connection with the management cluster** should only be required to write resources. For a dry run, no connection should be necessary. (However, an internet connection will be required to access public information from GitHub repositories.)
+
+- Configuration provided by the user should be **validated** against the apps's schema, and in case of validation errors, users should receive detailed feedback.
+
+- Users should decide whether to actually create the cluster, or print a manifest only.
+
+- A **cluster name must be provided by the user**. We won't generate a cluster name for the user.
+  - At Giant Swarm we decided around November 2023 that we want users to set workload cluster names and avoid auto-generated names ([roadmap#2999](https://github.com/giantswarm/roadmap/issues/2999)). In theory, we could still offer an option (a flag) to auto-generate the name. However, this would require knowledge of the cluster-PROVIDER app schema in the command logic. For example, as of cluster-aws v0.73.0, the cluster name is set in the property .global.metadata.name. Although we are working towards alignment across providers, this may be different in other versions or other provider apps.
+
+## Basic syntax
+
+The command should be named `create cluster`, hence the entire command, without any flags, would be:
+
+    kubectl gs create cluster
+
+The most important flags would be:
+
+- `--provider`: (required) provider name. For example: capa, capvcd, capz, eks.
+  - TODO: How do we restrict and validate the provider list per installation?
+- `--config`: Path to a config file (optional). Can be used multiple times. See "Configuration layering" below. As an alternative, or in addition, configuration can be passed via standard input.
+- `--app-version`: the cluster-PROVIDER app version to use. If not specified, the latest version is used.
+- `--dry-run`: only validate the configuration and print the resulting manifest.
+- `--output`: file path to write the result to. If not specified, resources are written to the management cluster. The option `STDOUT` can be used to print the result to the console.
+- `--format`: This flag can be used to specify the format of the output.
+  - Default is `manifest` for a Kubernetes manifest that includes an App resources and one or several ConfigMap resources.
+  - `config` for merged configuration YAML.
+  - `command` for a self-contained representation of the command just executed, as a one-liner.
+- `--set`: Override a configuration value. This flag can be used multiple times. The format is `path.to.property=value`.
+  - Note: `--set` can only be used with scalar values (string, number, boolean, integer, etc.). It cannot be used to override an array or object.
+
+TODO:
+
+- How should the resource **namespace** be specified? It could be derived from the organization name included in the cluster config, however that would require logical understanding of the config, which may differ over time and between providers.
```

Comment:
But all providers run cluster resources on the org namespace, could we just use that?

Reply:
We can assume in the CLI command implementation that the org name is specified in `.global.metadata.organization` and derive the namespace from that. However, if the cluster app schema changes and the org name field moves into a different place, this implementation will break and have to be updated, with the usual quirks of maintaining backward compatibility and supporting several versions.

Reply:
can we make a test to find in case it breaks?

Reply:
Hmm, from what I'm reading here, it seems we have an org field separate from the org namespace? Why? This feels wrong to me to duplicate and by that generate implementation blockers for moving and changing namespaces. K8s resources should not have dependencies on their namespace.

Reply:
> if the cluster app schema changes and the org name field moves into a different place, this implementation will break

Isn't it also true for all the other flags?

Reply:
> Hmm, from what I'm reading here, it seems we have an org field separate from the org namespace?

Well yes, it's part of the cluster config schema. We might get rid of it (outside the scope of this RFC) but then we still need a solution to specify the org namespace.

>> if the cluster app schema changes and the org name field moves into a different place, this implementation will break

> Isn't it also true for all the other flags?

Which other flags? There should not be any duplication between flags and cluster config. There is the `--set` flag, which takes the full property path and the value, so the command implementation does not need knowledge about the schema. Example: `--set global.metadata.name=test02`.

Reply:
Since it's so conventional with `kubectl` to use the `--namespace/-n` flag, probably the most straightforward would be to just to ask for that flag. However, there are some caveats:

- If not specified, the typical (expected by users) default namespace value is `default`. This means a user could accidentally try to create a cluster in that namespace (preventing this would be a matter of permissions or admission control and outside of the CLI's jurisdiction). It also means that a different defaulting behaviour to set some org-namespace would be specific to this command and unexpected. I am not in favour of this.

- As long as there is [`.global.metadata.organization` in the schema as a required property](https://github.com/giantswarm/cluster/blob/v0.29.0/helm/cluster/values.schema.json#L1450), we would ask for redundant input. I am also not in favour of this.

Reply:
ok, not part of this RFC to fix this, but we should definitely see that we revisit this org field, I'm still a bit surprised we duplicate info like this and tie things into our proprietary business logic

Reply:
Marking this as resolved. True, the metadata field in the schema should be deprecated.